dash/dash.py

```diff
@@ -108,8 +108,13 @@ def __init__(
             components_cache_max_age=None,
             show_undo_redo=False,
             plugins=None,
+            defer_app=False,
```

Comment:
I feel like our kwargs are getting a bit out of hand - can we accomplish this without adding a new one? Like, perhaps switch the `server` kwarg default to `'new'`, and that creates a new `Flask` app rather than looking for falsy like we do now, but you can pass in `None` (or really anything that's neither `'new'` nor an instance of `Flask`) and we'll treat that like your `defer_app=True`.

@chriddyp does this make sense to you or is my aversion to new kwargs misplaced?

Either way it's well past time that we make a docstring for this class describing all the kwargs! Not your problem @TMiguelT, I'll take care of that.

Reply:
The reason I avoided changing the meaning of any kwargs is because that would break backwards compatibility. However, we could still have a `server='defer'` mode that would reduce the number of kwargs and not break anything.

Reply:
Right, I appreciate your effort to maintain compatibility. This would be a pretty minor breakage though - only if someone explicitly provides a falsy `server` and expects it to generate the automatic one. I won't claim *nobody* has a use case for that, but I certainly can't think of a common one. And if we're going to be making breaking changes, now is the time - the next release will be dash 1.0.

Reply:
I conceptually agree that `server` seems like a decent place for this, given that they have very similar intentions.

Reply:
Okay, so how about:
* `server=True`: Create new server
* `server=False`: Defer creating server
* `server=Flask()`: Use existing flask server